awsshell/selectmenu.py

```diff
@@ -0,0 +1,264 @@
+        self.width = max(get_cwidth(o) for o in self._options) + 2
+        self.width = max(self.width, self.MIN_WIDTH)
+        self.height = len(options)
+        self._selection = None
+
+    def get_selection(self):
+        """Return the currently selected option, if there is one."""
+        if self._selection is not None:
+            return self._options[self._selection]
+        else:
+            return None
+
+    def select_down(self, event=None):
+        """Cycle down the list of options by one."""
+        if self._selection is not None:
+            self._selection = (self._selection + 1) % self.height
+        else:
+            self._selection = 0
+        self._insert_text(event)
+
+    def select_up(self, event=None):
+        """Cycle up the list of options by one."""
+        if self._selection is not None:
+            self._selection -= 1
+            if self._selection < 0:
+                self._selection = self.height - 1
+        else:
+            self._selection = self.height - 1
+        self._insert_text(event)
+
+    def _insert_text(self, event):
+        if event is not None:
+            event.current_buffer.document = Document(self.get_selection())
+
+    def preferred_width(self, cli, max_available_width):
+        """Return the preferred width of this UIControl."""
+        return self.width
+
+    def preferred_height(self, cli, width, max_available_height, wrap_lines):
+        """Return the preferred height of this UIControl."""
+        return self.height
+
+    def create_content(self, cli, width, height):
+        """Generate the UIContent for this control.
+
+        Create a get_line function that returns how each line of this control
+        should be rendered.
+        """
+        def get_line(i):
+            c = self._options[i]
+            is_current = (i == self._selection)
+            # Render the line as array of tokens for highlighting
+            return self._get_menu_item_tokens(c, is_current)
+
+        return UIContent(
+            get_line=get_line,
+            line_count=self.height,
+            default_char=Char(' ', self.token),
+            cursor_position=Point(x=0, y=self._selection or 0)
+        )
+
+    def _get_menu_item_tokens(self, option, is_current):
+        """Given an option generate the proper token list for highlighting."""
+        # highlight the current selection with a different token
+        if is_current:
+            token = self.token.Completion.Current
+        else:
+            token = self.token.Completion
+        # pad all lines to the same width
+        padding = ' ' * (self.width - len(option))
+        return [(token, ' %s%s ' % (option, padding))]
+
+
+def create_select_menu_layout(msg, menu_control,
+                              show_meta=False,
+                              reserve_space_for_menu=True):
+    """Construct a layout for the given message and menu control."""
+    def get_prompt_tokens(_):
```

Comment:
This is a callback that's called when a `TokenListControl` object attempts to create content to be rendered. This function is passed the `CommandLineInterface` object that contains the `TokenListControl`. This is what defines how the prompt should be rendered and it does not depend on the CLI, thus the only argument passed in is ignored and is why I made the decision to use `_` instead of something else like `cli` as the variable name.

For reference prompt_toolkit defined this as:
```
if get_prompt_tokens is None:
        get_prompt_tokens = lambda _: [(Token.Prompt, message)]
```